ensime_shared/editor.py

```diff
@@ -0,0 +1,384 @@
+    # ftplugin/scala/ensime.vim, custom/dotted filetypes & syntax
+    def initialize(self):
+        """Sets up initial ensime-vim editor settings."""
+        self._vim.options['updatetime'] = 1000
+
+        # TODO: This seems wrong, the user setting value is never used anywhere.
+        if 'EnErrorStyle' not in self._vim.vars:
+            self._vim.vars['EnErrorStyle'] = 'EnError'
+        self._vim.command('highlight EnErrorStyle ctermbg=red gui=underline')
+
+        # TODO: this SHOULD be a buffer-local setting only, and since it should
+        # apply to all Scala files, ftplugin is the ideal place to set it. I'm
+        # not even sure how this is currently working when only set once.
+        self._vim.command('set omnifunc=EnCompleteFunc')
+
+        # TODO: custom filetype ftplugin
+        self._vim.command(
+            'autocmd FileType package_info nnoremap <buffer> <Space> :call EnPackageDecl()<CR>')
+        self._vim.command('autocmd FileType package_info setlocal splitright')
+
+    # TODO: make this a R/W property?
+    def cursor(self):
+        """Get the cursor position in the current window as a ``(row, column)``
+        tuple.
+        """
+        return self._vim.current.window.cursor
+
+    def set_cursor(self, row, col):
+        """Set cursor position to given row and column in the current window."""
+        self._vim.current.window.cursor = (row, col)
+
+    # TODO: this can be replaced with a ``getpos()`` expression
+    def start_end_pos(self):
+        """Return start and end positions of the cursor respectively."""
+        self._vim.command('normal e')
+        end = self.cursor()
+        self._vim.command('normal b')
+        beg = self.cursor()
+        return beg, end
+
+    def path(self):
+        """Return the current path."""
+        return self._vim.current.buffer.name
+
+    def get_file_content(self):
+        """Get content of file."""
+        return "\n".join(self._vim.current.buffer)
+
+    # This is used only once, maybe just make a higher-level API or inline it
+    def width(self):
+        """Return the width of the window."""
+        return self._vim.current.window.width
+
+    def ask_input(self, prompt):
+        """Prompt user for input and return the entered value."""
+        self._vim.command('call inputsave()')
+        self._vim.command('let user_input = input("{} ")'.format(prompt))
+        self._vim.command('call inputrestore()')
+        response = self._vim.eval('user_input')
+        self._vim.command('unlet user_input')
+        return response
+
+    def to_quickfix_item(self, file_name, line_number, message, tpe):
+        return {"filename": file_name,
+                "lnum": line_number,
+                "text": message,
+                "type": tpe}
+
+    def write_quickfix_list(self, qflist):
+        self._vim.command('call setqflist({!s})'.format(qflist))
+        self._vim.command('copen')
+
+    def lazy_display_error(self, filename):
```

Comment:
Is it for displaying compilation errors? Isn't this functionality provided by Syntastic already?